### PR TITLE
test: modernize tests

### PR DESCRIPTION
**Pull request by angeloashmore**
Resolves: N/A

### Description

This PR reimplements all of `@prismicio/client`'s tests with a modern approach. The new tests use the real Prismic API and minimizes network mocking. The new tests are more like E2E tests than integration tests.

This rewrite results in much shorter tests (~27.5k lines -> ~7.5k lines total), which should be easier to maintain. Using the real API also means we will be more aware of compatability issues.

Claude Code was used to help convert tests, but all lines were manually reviewed and edited as needed.

Notable changes:
- Uses `@prismicio/e2e-tests-utils` to set up a shared test repository per run.
- Uses the real Prismic API to test the client's behavior; `@prismicio/mock` is not used.
- `fetch` responses are sometimes mocked to test edge cases.

> [!IMPORTANT]
> This PR includes a few client fixes caught while reimplementing the tests. The changes are annotated in the file view.

**This PR only adds tests.** The new tests are in `test2/`. The old tests remain in the codebase under `test/` and are ignored. The type tests in `test/` are still used.

After merging this PR, a separate PR will be opened, including:
- Moving the type tests to `test2/`.
- Removing the current `test/`.
- Renaming `test2/` to `test/`.
- Removing unused dependencies.

### Checklist

- [ ] A comprehensive Linear ticket, providing sufficient context and details to facilitate the review of the PR, is linked to the PR.
- [ ] If my changes require tests, I added them.
- [ ] If my changes affect backward compatibility, it has been discussed.
- [ ] If my changes require an update to the CONTRIBUTING.md guide, I updated it.

### How to QA [^1]

[^1]:
	Please use these labels when submitting a review:
	:question: #ask:&ensp;Ask a question.
	:bulb: #idea:&ensp;Suggest an idea.
	:warning: #issue:&ensp;Strongly suggest a change.
	:tada: #nice:&ensp;Share a compliment.

### CODE REVIEW

File: src/buildQueryURL.ts [hunk 1]
```diff
@@ -326,7 +326,10 @@ export const buildQueryURL = (
 ): string => {
 	const { filters, predicates, ...params } = args
 
-	const url = new URL(`documents/search`, `${endpoint}/`)
+	if (!endpoint.endsWith("/")) {
+		endpoint += "/"
+	}
```
Comment (angeloashmore) [inline src/buildQueryURL.ts]: Trailing slash should have always only been added if it wasn't already there.

File: vite.config.ts [hunk 2]
```diff
@@ -12,14 +13,17 @@ export default defineConfig({
 		},
 	},
 	test: {
+		include: [".\/test2\/**/*.test.ts"],
+		globalSetup: ["./test2/setup.global.ts"],
+		setupFiles: ["./test2/setup.ts"],
 		typecheck: {
 			enabled: true,
 		},
+		retry: 1,
```
Comment (angeloashmore) [inline vite.config.ts]: Very few tests are flaky. Retrying once prevents virtually all flaky tests.

File: test2/asDate.test.ts
Comment (lihbr) [inline test2/asDate.test.ts]: 💡 #idea: I believe we should still `helpers-`-prefix the helper tests as not all helpers are `as`-prefixed (`isFilled`, `mapSliceZone`, etc.)

File: test2/setup.global.ts [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+	})
+	const french2 = await createDocument(repo, model.id, {
+		locale: "fr-fr",
+		tags: ["bar"],
+		routes,
+	})
+	const frenchSingle = await createDocument(repo, singleModel.id, {
+		locale: "fr-fr",
+		routes,
+	})
```
Comment (lihbr) [inline test2/setup.global.ts]: 💡 #idea: Have we considered using `Promise.all` here? It's quite a slow part of the process, I feel 🤔

If publication is the issue, I was quite successful at creating many documents by creating them in a release and then publishing the release to publish them all in one go.

**Review by lihbr — COMMENTED** [review]
Hey, this looks really good to me so far (about halfway through), I've shared some ideas for now, will finish the review tomorrow ☺️

Tests run fine on my end, but those are quite flaky; maybe we should consider a retry on them (I guess it's mainly v8 warmup related):

Image

<details>
<summary>Failing tests list</summary>

```
 FAIL  test2/client-getRefByID.test.ts > returns ref with matching ID
 FAIL  test2/client-getRefByID.test.ts > returns ref with matching ID
 FAIL  test2/client-getRefByID.test.ts > shares concurrent equivalent network requests
 FAIL  test2/client-getRefByID.test.ts > shares concurrent equivalent network requests
 FAIL  test2/client-getRefByLabel.test.ts > returns ref with matching label
 FAIL  test2/client-getRefByLabel.test.ts > returns ref with matching label
 FAIL  test2/client-getRefByLabel.test.ts > shares concurrent equivalent network requests
 FAIL  test2/client-getRefByLabel.test.ts > shares concurrent equivalent network requests
 FAIL  test2/client-getReleaseByID.test.ts > returns release with matching ID
 FAIL  test2/client-getReleaseByID.test.ts > returns release with matching ID
 FAIL  test2/client-getReleaseByID.test.ts > shares concurrent equivalent network requests
 FAIL  test2/client-getReleaseByID.test.ts > shares concurrent equivalent network requests
 FAIL  test2/client-getReleaseByLabel.test.ts > returns release with matching label
 FAIL  test2/client-getReleaseByLabel.test.ts > returns release with matching label
 FAIL  test2/client-getReleaseByLabel.test.ts > shares concurrent equivalent network requests
 FAIL  test2/client-getReleaseByLabel.test.ts > shares concurrent equivalent network requests
 FAIL  test2/client-queryContentFromReleaseByID.test.ts > fetches content from a release
 FAIL  test2/client-queryContentFromReleaseByID.test.ts > fetches content from a release
 FAIL  test2/client-queryContentFromReleaseByID.test.ts > uses the cached release ref within the refs TTL
 FAIL  test2/client-queryContentFromReleaseByID.test.ts > uses the cached release ref within the refs TTL
 FAIL  test2/client-queryContentFromReleaseByLabel.test.ts > fetches content from a release
 FAIL  test2/client-queryContentFromReleaseByLabel.test.ts > fetches content from a release
 FAIL  test2/client-queryContentFromReleaseByLabel.test.ts > uses the cached release ref within the refs TTL
 FAIL  test2/client-queryContentFromReleaseByLabel.test.ts > uses the cached release ref within the refs TTL
 FAIL  test2/client.test.ts > 'getMasterRef' > supports fetch options
 FAIL  test2/client.test.ts > 'getMasterRef' > supports fetch options
 FAIL  test2/client.test.ts > 'getMasterRef' > supports default fetch options
 FAIL  test2/client.test.ts > 'getMasterRef' > supports default fetch options
 FAIL  test2/client.test.ts > 'getMasterRef' > supports signal
 FAIL  test2/client.test.ts > 'getMasterRef' > supports signal
 FAIL  test2/client.test.ts > 'getRefByID' > supports fetch options
 FAIL  test2/client.test.ts > 'getRefByID' > supports fetch options
 FAIL  test2/client.test.ts > 'getRefByID' > supports default fetch options
 FAIL  test2/client.test.ts > 'getRefByID' > supports default fetch options
 FAIL  test2/client.test.ts > 'getRefByID' > supports signal
 FAIL  test2/client.test.ts > 'getRefByID' > supports signal
 FAIL  test2/client.test.ts > 'getRefByLabel' > supports fetch options
 FAIL  test2/client.test.ts > 'getRefByLabel' > supports fetch options
 FAIL  test2/client.test.ts > 'getRefByLabel' > supports default fetch options
 FAIL  test2/client.test.ts > 'getRefByLabel' > supports default fetch options
 FAIL  test2/client.test.ts > 'getRefByLabel' > supports signal
 FAIL  test2/client.test.ts > 'getRefByLabel' > supports signal
 FAIL  test2/client.test.ts > 'getRefs' > supports fetch options
 FAIL  test2/client.test.ts > 'getRefs' > supports fetch options
 FAIL  test2/client.test.ts > 'getRefs' > supports default fetch options
 FAIL  test2/client.test.ts > 'getRefs' > supports default fetch options
 FAIL  test2/client.test.ts > 'getRefs' > supports signal
 FAIL  test2/client.test.ts > 'getRefs' > supports signal
 FAIL  test2/client.test.ts > 'getReleaseByID' > supports fetch options
 FAIL  test2/client.test.ts > 'getReleaseByID' > supports fetch options
 FAIL  test2/client.test.ts > 'getReleaseByID' > supports default fetch options
 FAIL  test2/client.test.ts > 'getReleaseByID' > supports default fetch options
 FAIL  test2/client.test.ts > 'getReleaseByID' > supports signal
 FAIL  test2/client.test.ts > 'getReleaseByID' > supports signal
 FAIL  test2/client.test.ts > 'getReleaseByLabel' > supports fetch options
 FAIL  test2/client.test.ts > 'getReleaseByLabel' > supports fetch options
 FAIL  test2/client.test.ts > 'getReleaseByLabel' > supports default fetch options
 FAIL  test2/client.test.ts > 'getReleaseByLabel' > supports default fetch options
 FAIL  test2/client.test.ts > 'getReleaseByLabel' > supports signal
 FAIL  test2/client.test.ts > 'getReleaseByLabel' > supports signal
 FAIL  test2/client.test.ts > 'getReleases' > supports fetch options
 FAIL  test2/client.test.ts > 'getReleases' > supports fetch options
 FAIL  test2/client.test.ts > 'getReleases' > supports default fetch options
 FAIL  test2/client.test.ts > 'getReleases' > supports default fetch options
 FAIL  test2/client.test.ts > 'getReleases' > supports signal
 FAIL  test2/client.test.ts > 'getReleases' > supports signal
 FAIL  test2/client.test.ts > 'getTags' > supports fetch options
 FAIL  test2/client.test.ts > 'getTags' > supports fetch options
 FAIL  test2/client.test.ts > 'getTags' > supports default fetch options
 FAIL  test2/client.test.ts > 'getTags' > supports default fetch options
 FAIL  test2/client.test.ts > 'getTags' > supports signal
 FAIL  test2/client.test.ts > 'getTags' > supports signal
 FAIL  test2/client.test.ts > 'getRepository' > supports fetch options
 FAIL  test2/client.test.ts > 'getRepository' > supports fetch options
 FAIL  test2/client.test.ts > 'getRepository' > supports default fetch options
 FAIL  test2/client.test.ts > 'getRepository' > supports default fetch options
 FAIL  test2/client.test.ts > 'getRepository' > supports signal
 FAIL  test2/client.test.ts > 'getRepository' > supports signal
 FAIL  test2/client.test.ts > 'buildQueryURL' > supports fetch options
 FAIL  test2/client.test.ts > 'buildQueryURL' > supports fetch options
 FAIL  test2/client.test.ts > 'buildQueryURL' > supports default fetch options
 FAIL  test2/client.test.ts > 'buildQueryURL' > supports default fetch options
 FAIL  test2/client.test.ts > 'buildQueryURL' > supports signal
 FAIL  test2/client.test.ts > 'buildQueryURL' > supports signal
 FAIL  test2/client.test.ts > 'resolvePreviewURL' > supports fetch options
 FAIL  test2/client.test.ts > 'resolvePreviewURL' > supports fetch options
 FAIL  test2/client.test.ts > 'resolvePreviewURL' > supports default fetch options
 FAIL  test2/client.test.ts > 'resolvePreviewURL' > supports default fetch options
 FAIL  test2/client.test.ts > 'resolvePreviewURL' > supports signal
 FAIL  test2/client.test.ts > 'resolvePreviewURL' > supports signal
```

</details>

File: src/WriteClient.ts [hunk 4]
```diff
@@ -372,7 +372,7 @@ export class WriteClient<
 	): Promise<void> {
 		// Resolve master locale
 		const repository = await this.getRepository(fetchParams)
-		const masterLocale = repository.languages.find((lang) => lang.is_master)!.id
+		const masterLocale = repository.languages[0].id
```
[the remark this answers is not in the document]
Comment (angeloashmore) [reply src/WriteClient.ts]: I'm not sure what determines if `is_master` is present. It's missing in some of my older repos as well as the new test repos generated in `setup.global.ts`.

Example: https://prismic.io/explorer?request=%7B%22repository%22%3A%7B%22id%22%3A%22sveltekit-starter-prismic-multi-page%22%7D%2C%22type%22%3A%22getRepository%22%2C%22options%22%3A%7B%7D%2C%22pagination%22%3A%7B%7D%2C%22advanced%22%3Afalse%2C%22filters%22%3A%5B%5D%2C%22orderings%22%3A%5B%5D%7D&source=sharing

See this related thread: https://prismic-team.slack.com/archives/C02JLEV98G3/p1704271780178709?thread_ts=1704201728.851369&cid=C02JLEV98G3

I found this bug while writing the migration tests since it checks for the master lang. It wasn't able to find it using the old implementation.

If you think we need to strengthen the master language selection, we can check for `is_master` and fallback to `[0]`, but it seems like we have a guarantee that the first language will be the master.

File: test2/asImagePixelDensitySrcSet.test.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
+import { it } from "./it"
+
+import { asImagePixelDensitySrcSet } from "../src"
+
+const field = {
```
[the remark this answers is not in the document]
Comment (angeloashmore) [reply test2/asImagePixelDensitySrcSet.test.ts]: Thanks, good catch. I added an explicit type to all tests with `const field`.

File: test2/setup.global.ts (hunk 3 above)
Comment (angeloashmore) [reply test2/setup.global.ts]: Promise.all worked locally and was much faster, as expected. It fails almost always in CI, however ([example](https://github.com/prismicio/prismic-client/actions/runs/19114304525/job/54619377911)).

Using a release with a custom `waitFor` is almost as fast, but much more reliable.

The total runtime in CI is nearly identical to before the refactor, but local setup goes from ~40s to ~15s locally. Still a win, IMO.

File: test2/helpers-asDate.test.ts
Comment (angeloashmore) [reply test2/helpers-asDate.test.ts]: Done! In the upcoming refactor, I think we'll remove the `helpers-` prefix and the `src/helpers` directory to more closely match the public API. Not decided yet; we'll need to play around with it.

File: test2/setup.ts [hunk 6]
```diff
@@ -0,0 +1,351 @@
+import type { MockInstance } from "vitest"
+import { afterEach, beforeEach, expect, vi } from "vitest"
+
+import { deepEqual } from "node:assert/strict"
```
[the remark this answers is not in the document]
Comment (angeloashmore) [reply test2/setup.ts]: Lots of nice APIs now!

File: test2/setup.ts [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+
+		return {
+			pass: actual === expected,
+			message: () =>
+				`Client fetched Content API ${actual} time${actual === 1 ? "" : "s"}.`,
+			actual,
+			expected,
+		}
+	},
+	toHaveFetchedRepo(client: unknown, expectedParams, expectedInit = {}) {
```
[the remark this answers is not in the document]
Comment (angeloashmore) [reply test2/setup.ts]: You're right; I only shortened it so we could fit more code on a line. I think developers generally know "repo" as "repository" (e.g. "Git repo"), but it's simpler to spell everything out fully.

I'll go through the files and update `repo`, `repos`, and any other occurrences.

**Comment by angeloashmore** [comment]
Thanks for the review, @lihbr. I'll merge and open a separate PR to remove the old tests.
